### ISSUE TITLE
TestIntegrationFinalPath failure

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#36522]
TestIntegrationFinalPath failed in CMSSW_12_3_X_2021-12-15-2300 slc7_amd64_gcc900 with
```
/data/cmsbld/jenkins/workspace/ib-run-qa/CMSSW_12_3_X_2021-12-15-2300/tmp/slc7_amd64_gcc900 /data/cmsbld/jenkins/workspace/ib-run-qa/CMSSW_12_3_X_2021-12-15-2300
1,7c1
< found thing '' TEST
< found thing '' TEST
< found thing '' TEST
< found thing 'beginLumi' TEST
< found thing 'endLumi' TEST
< found thing 'beginRun' TEST
< found thing 'endRun' TEST
---
> Binary file finalpath.log matches
differences for test_finalpath_cfg.py --path --task --schedule: status 1
=== Log file ===
=== End log file ===

---> test TestIntegrationFinalPath had ERRORS
```
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_3_X_2021-12-15-2300/unitTestLogs/FWCore/Integration#/

**Comment by makortel** [comment]
assign core

**Comment by makortel** [comment]
So something caused `grep` to interpret `finalpath.log` as a binary file?
https://github.com/cms-sw/cmssw/blob/4d5c91a05629d9124806e310501fac8c303c84c2/FWCore/Integration/test/test_finalpath.sh#L60-L61

**Comment by makortel** [comment]
Occurred again in CMSSW_12_3_X_2021-12-24-2300

**Comment by dan131riley** [comment]
`^@` (control-@) is how browsers and pagers (and some editors) conventionally render the `nul` character, so we're writing out a `\0` somewhere.

**Comment by Dr15Jones** [comment]
In order to help diagnose the underlying problem, the actual message should have been
```
%MSG-i FoundProduct:  PostGlobalEndRun 25-Jan-2022 16:47:03 UTC PostGlobalEndRun
found thing 'beginRun' TEST
%MSG
%MSG-i FoundProduct:  PostGlobalEndRun 25-Jan-2022 16:47:03 UTC PostGlobalEndRun
found thing 'endRun' TEST
%MSG
```

**Comment by Dr15Jones** [comment]
Intersting enough, the `\0` Dan points out is exactly where `\0` would be if one were to write `PostGlobalEndRun` and then stick the `\0` at the end.

The item holding the information in question is a `std::string_view` which doesn't have to use `\0` as its end condition.

**Comment by Dr15Jones** [comment]
The incorrect messages seems to contain `MessageDrop::runEvent` from the `preStreamEndRun` transition

https://github.com/cms-sw/cmssw/blob/ec200b2e6d0213cdf868308a7a836be8912c4cbc/FWCore/MessageService/plugins/MessageLogger.cc#L786-L790

**Comment by Dr15Jones** [comment]
I was also able to reproduce the problem. The full log file is
```
['cmsRun', '/root/buildArea/CMSSW_12_3_0_pre2/src/FWCore/Integration/test/test_finalpath_cfg.py', '--', '--path', '--task']
adding Path
%MSG-i ThreadStreamSetup:  (NoModuleName) 25-Jan-2022 17:23:33 UTC pre-events
setting # threads 3
setting # streams 1
%MSG
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 25-Jan-2022 17:23:35.156 UTC
%MSG-i FoundProduct:  GetProductCheckerOutputModule:out  25-Jan-2022 17:23:35 UTC Run: 1 Event: 1
found thing '' TEST
%MSG
Begin processing the 2nd record. Run 1, Event 2, LumiSection 1 on stream 0 at 25-Jan-2022 17:23:35.157 UTC
%MSG-i FoundProduct:  GetProductCheckerOutputModule:out  25-Jan-2022 17:23:35 UTC Run: 1 Event: 2
found thing '' TEST
%MSG
Begin processing the 3rd record. Run 1, Event 3, LumiSection 1 on stream 0 at 25-Jan-2022 17:23:35.157 UTC
%MSG-i FoundProduct:  GetProductCheckerOutputModule:out  25-Jan-2022 17:23:35 UTC Run: 1 Event: 3
found thing '' TEST
%MSG
%MSG-i FoundProduct:  PostGlobalEndLumi 25-Jan-2022 17:23:35 UTC  PostGlobalEndLumi
found thing 'beginLumi' TEST
%MSG
%MSG-i FoundProduct:  PostGlobalEndLumi 25-Jan-2022 17:23:35 UTC  PostGlobalEndLumi
found thing 'endLumi' TEST
%MSG
%MSG-i FoundProduct:  PreStreamEndLumi 25-Jan-2022 17:23:35 UTC  End Run: 1 Stream: 0^@: 0
found thing 'beginRun' TEST
%MSG
%MSG-i FoundProduct:  PreStreamEndLumi 25-Jan-2022 17:23:35 UTC  End Run: 1 Stream: 0^@: 0
found thing 'endRun' TEST
%MSG
```

**Comment by Dr15Jones** [comment]
I think I now understand the origin of the problem. The MessageLogger does not listen to the ES module nor the OutputModule write signals. Therefore the context it uses for such messages is just whatever context was used for the last known transition to be run on that thread.

**Comment by makortel** [comment]
Thanks @Dr15Jones. Should we make MessageLogger to listen ES module and OutputModule write signals then?

Do these observations also imply we're using `std::string_view` in some place within MessageLogger where we shouldn't?

**Comment by Dr15Jones** [comment]
> Thanks @Dr15Jones. Should we make MessageLogger to listen ES module and OutputModule write signals then?

I'm working on that

> Do these observations also imply we're using std::string_view in some place within MessageLogger where we shouldn't?

Not so far.

**Comment by makortel** [comment]
+1

**Comment by perrotta** [comment]
@makortel I imagine that since you signed this issue, it can also get closed: however, I'd let you do it, just in case you wanted to have it still open for other purposes.

**Comment by makortel** [comment]
@perrotta It was already closed by the merge of https://github.com/cms-sw/cmssw/pull/36801

**Comment by perrotta** [comment]
> @perrotta It was already closed by the merge of #36801

You are right Matti, of course!
Sorry for the noise, I didn't realize it.
